<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Chapter-13" data-toc-modified-id="Chapter-13-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Chapter 13</a></span><ul class="toc-item"><li><span><a href="#Example-Multilevel-Tadpoles" data-toc-modified-id="Example-Multilevel-Tadpoles-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Example Multilevel Tadpoles</a></span></li><li><span><a href="#Varying-effects-and-the-underfitting/overfitting-trade-off" data-toc-modified-id="Varying-effects-and-the-underfitting/overfitting-trade-off-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Varying effects and the underfitting/overfitting trade-off</a></span><ul class="toc-item"><li><span><a href="#The-model" data-toc-modified-id="The-model-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>The model</a></span></li><li><span><a href="#Assign-values-to-the-parameters" data-toc-modified-id="Assign-values-to-the-parameters-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Assign values to the parameters</a></span></li><li><span><a href="#Simulate-Survivors" data-toc-modified-id="Simulate-Survivors-1.2.3"><span class="toc-item-num">1.2.3&nbsp;&nbsp;</span>Simulate Survivors</a></span></li><li><span><a href="#Compute-the-no-pooling-estimates" data-toc-modified-id="Compute-the-no-pooling-estimates-1.2.4"><span class="toc-item-num">1.2.4&nbsp;&nbsp;</span>Compute the no-pooling estimates</a></span></li><li><span><a href="#Compute-the-partial-pooling-estimates" data-toc-modified-id="Compute-the-partial-pooling-estimates-1.2.5"><span class="toc-item-num">1.2.5&nbsp;&nbsp;</span>Compute the partial-pooling estimates</a></span></li></ul></li><li><span><a href="#More-than-one-type-of-cluster" data-toc-modified-id="More-than-one-type-of-cluster-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>More than one type of cluster</a></span><ul class="toc-item"><li><span><a href="#Multilevel-chimpanzees" data-toc-modified-id="Multilevel-chimpanzees-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Multilevel chimpanzees</a></span></li><li><span><a href="#Even-more-clusters" data-toc-modified-id="Even-more-clusters-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>Even more clusters</a></span></li></ul></li><li><span><a href="#Divergent-transitions-and-non-centered-priors" data-toc-modified-id="Divergent-transitions-and-non-centered-priors-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Divergent transitions and non-centered priors</a></span><ul class="toc-item"><li><span><a href="#Adjust-target-acceptance" data-toc-modified-id="Adjust-target-acceptance-1.4.1"><span class="toc-item-num">1.4.1&nbsp;&nbsp;</span>Adjust target acceptance</a></span></li><li><span><a href="#Reparameterization" data-toc-modified-id="Reparameterization-1.4.2"><span class="toc-item-num">1.4.2&nbsp;&nbsp;</span>Reparameterization</a></span></li></ul></li><li><span><a href="#Multilevel-posterior-predictions" data-toc-modified-id="Multilevel-posterior-predictions-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Multilevel posterior predictions</a></span><ul class="toc-item"><li><span><a href="#Posterior-prediction-for-same-clusters." data-toc-modified-id="Posterior-prediction-for-same-clusters.-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Posterior prediction for same clusters.</a></span></li><li><span><a href="#Posterior-prediction-for-new-clusters." data-toc-modified-id="Posterior-prediction-for-new-clusters.-1.5.2"><span class="toc-item-num">1.5.2&nbsp;&nbsp;</span>Posterior prediction for new clusters.</a></span></li></ul></li><li><span><a href="#Practice-Problems" data-toc-modified-id="Practice-Problems-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Practice Problems</a></span></li></ul></li><li><span><a href="#---" data-toc-modified-id="----2"><span class="toc-item-num">2&nbsp;&nbsp;</span>---</a></span></li><li><span><a href="#Misc" data-toc-modified-id="Misc-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Misc</a></span><ul class="toc-item"><li><span><a href="#How-link-works" data-toc-modified-id="How-link-works-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>How link works</a></span></li></ul></li></ul></div>

In [1]:
library(tidyverse)

# loads experimental branch of statistical rethinking pkg
library(rethinking, lib = "C:/Users/bings/Documents/R/test-library")

-- Attaching packages --------------------------------------- tidyverse 1.2.1 --
v ggplot2 3.1.0       v purrr   0.3.0  
v tibble  2.0.1       v dplyr   0.8.0.1
v tidyr   0.8.2       v stringr 1.4.0  
v readr   1.3.1       v forcats 0.4.0  
Warning message:
"package 'tibble' was built under R version 3.5.2"Warning message:
"package 'readr' was built under R version 3.5.2"Warning message:
"package 'purrr' was built under R version 3.5.2"Warning message:
"package 'dplyr' was built under R version 3.5.2"Warning message:
"package 'stringr' was built under R version 3.5.2"Warning message:
"package 'forcats' was built under R version 3.5.2"-- Conflicts ------------------------------------------ tidyverse_conflicts() --
x dplyr::filter() masks stats::filter()
x dplyr::lag()    masks stats::lag()
Loading required package: rstan
Loading required package: StanHeaders
Warning message:
"package 'StanHeaders' was built under R version 3.5.2"rstan (Version 2.18.2, GitRev: 2e1f913d3ca3)
For execution

# Chapter 13

## Example Multilevel Tadpoles

##### **<font color = 'purple' style='background-color: lightyellow'>Here is a model for predicting tadpole mortality in each tank, using the regularizing priors of earlier chapters:</font>**

<img src='pics/SR_r31.png' alt='Drawing' style='width:525pt'/>

In [2]:
data(reedfrogs)
d <- reedfrogs
str(d)

'data.frame':	48 obs. of  5 variables:
 $ density : int  10 10 10 10 10 10 10 10 10 10 ...
 $ pred    : Factor w/ 2 levels "no","pred": 1 1 1 1 1 1 1 1 2 2 ...
 $ size    : Factor w/ 2 levels "big","small": 1 1 1 1 2 2 2 2 1 1 ...
 $ surv    : int  9 10 7 10 9 9 10 9 4 9 ...
 $ propsurv: num  0.9 1 0.7 1 0.9 0.9 1 0.9 0.4 0.9 ...


In [3]:
d %>% head

density pred size  surv propsurv
1 10      no   big    9   0.9     
2 10      no   big   10   1.0     
3 10      no   big    7   0.7     
4 10      no   big   10   1.0     
5 10      no   small  9   0.9     
6 10      no   small  9   0.9

In [4]:
## R code 13.2
# make the tank cluster variable
d$tank <- 1:nrow(d)

dat <- list(
    S = d$surv,
    N = d$density,
    tank = d$tank )

In [5]:
# approximate posterior
m13.1 <- ulam(
    alist(
        S ~ dbinom( N , p ) ,
        logit(p) <- a[tank] ,
        a[tank] ~ dnorm( 0 , 1.5 )
    ), data=dat , chains=4 , log_lik=TRUE )


SAMPLING FOR MODEL 'cb3dfdd4abafffee16ea86593d6ba62d' NOW (CHAIN 1).
Chain 1: 
Chain 1: Gradient evaluation took 0 seconds
Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 0 seconds.
Chain 1: Adjust your expectations accordingly!
Chain 1: 
Chain 1: 
Chain 1: Iteration:   1 / 1000 [  0%]  (Warmup)
Chain 1: Iteration: 100 / 1000 [ 10%]  (Warmup)
Chain 1: Iteration: 200 / 1000 [ 20%]  (Warmup)
Chain 1: Iteration: 300 / 1000 [ 30%]  (Warmup)
Chain 1: Iteration: 400 / 1000 [ 40%]  (Warmup)
Chain 1: Iteration: 500 / 1000 [ 50%]  (Warmup)
Chain 1: Iteration: 501 / 1000 [ 50%]  (Sampling)
Chain 1: Iteration: 600 / 1000 [ 60%]  (Sampling)
Chain 1: Iteration: 700 / 1000 [ 70%]  (Sampling)
Chain 1: Iteration: 800 / 1000 [ 80%]  (Sampling)
Chain 1: Iteration: 900 / 1000 [ 90%]  (Sampling)
Chain 1: Iteration: 1000 / 1000 [100%]  (Sampling)
Chain 1: 
Chain 1:  Elapsed Time: 0.16 seconds (Warm-up)
Chain 1:                0.136 seconds (Sampling)
Chain 1:                0.2

In [ ]:
precis(m13.1,depth=2)

##### **<font color = 'purple' style='background-color: lightyellow'>Now let’s do the multilevel model, which adaptively pools information across tanks</font>**

All that is required to enable adaptive pooling is to make the prior for the a parameters a function of some new parameters. Here is the multilevel model, in mathematical form, with the changes from the previous model highlighted in blue:

<img src='pics/SR_r32.png' alt='Drawing' style='width:525pt'/>

Notice that the prior for the tank intercepts is now a function of two parameters, $\bar{α}$ and σ. You can say $\bar{α}$ like “bar alpha.” The bar means average. These two parameters inside the prior is where the “multi” in multilevel arises

In [7]:
## R code 13.3
m13.2 <- ulam(
    alist(
        S ~ dbinom( N , p ) ,
        logit(p) <- a[tank] ,
        a[tank] ~ dnorm( a_bar , sigma ) ,
        a_bar ~ dnorm( 0 , 1.5 ) ,
        sigma ~ dexp( 1 )
    ), data=dat , chains=4 , log_lik=TRUE )

Warning message in file(con, "w"):
"cannot open file 'C:\Users\bings\AppData\Local\Temp\Rtmp8YnaDH\file2efc2d7d3dec': No such file or directory"

ERROR: Error in file(con, "w"): cannot open the connection


This model provides posterior distributions for 50 parameters: one overall sample intercept $\bar{α}$, the standard deviation among tanks σ, and then 48 per-tank intercepts. Let’s check WAIC though to see the effective number of parameters. We’ll compare the earlier model, m13.1, with the new multilevel model:

In [8]:
precis(m13.2,depth=2) %>% round(2)

ERROR: Error in precis(m13.2, depth = 2): object 'm13.2' not found


This model provides posterior distributions for 50 parameters: one overall sample intercept α¯, the standard deviation among tanks σ, and then 48 per-tank intercepts. Let’s check WAIC though to see the effective number of parameters. We’ll compare the earlier model, m13.1, with the new multilevel model:

In [9]:
## R code 13.4
compare( m13.1 , m13.2 ) %>% round(2)

ERROR: Error in compare(m13.1, m13.2): object 'm13.1' not found


**<font color = 'purple'>There are two facts to note here.  </font>** 

**<font color = 'purple'>First </font>**, **the multilevel model has only 22 effective parameters**. <font color = 'blue'>There are 28 fewer effective parameters than actual parameters, because the prior assigned to each intercept shrinks them all towards the mean $\bar{α}$</font>. 

In this case, the prior is reasonably strong. Check the mean of sigma with precis and you’ll see it’s around 1.6. <font color = 'blue'>This is a regularizing prior, like you’ve used in previous chapters, but now the amount of regularization has been learned from the data itself.</font>

* **Note** that there is still uncertainty about the regularization. So this model isn’t exactly the same as just assuming a regularizing prior with a constant standard deviation 1.6. Instead the intercepts for each tank average over the uncertainty in σ (and $\bar{α}$)

**<font color = 'purple'>Second </font>**, **notice that the multilevel model m13.2 has fewer effective parameters than the ordinary fixed model m13.1**. This is despite the fact that the ordinary model has fewer actual parameters, only 48 instead of 50. 

* The extra two parameters in the multilevel model allowed it to learn a more aggressive regularizing prior, to adaptively regularize. This resulted in a less flexible posterior and therefore fewer effective parameters.

In [10]:
## R code 13.5
# extract Stan samples
post <- extract.samples(m13.2)

# compute median intercept for each tank
# also transform to probability with logistic
d$propsurv.est <- logistic( apply( post$a , 2 , mean ) )

# display raw proportions surviving in each tank
plot( d$propsurv , ylim=c(0,1) , pch=16 , xaxt="n" ,
    xlab="tank" , ylab="proportion survival" , col=rangi2 )
axis( 1 , at=c(1,16,32,48) , labels=c(1,16,32,48) )

# overlay posterior means
points( d$propsurv.est )

# mark posterior mean probability across tanks
abline( h=mean(inv_logit(post$a_bar)) , lty=2 )

# draw vertical dividers between tank densities
abline( v=16.5 , lwd=0.5 )
abline( v=32.5 , lwd=0.5 )
text( 8 , 0 , "small tanks" )
text( 16+8 , 0 , "medium tanks" )
text( 32+8 , 0 , "large tanks" )


ERROR: Error in extract.samples(m13.2): object 'm13.2' not found


In [ ]:
## R code 13.6
# show first 100 populations in the posterior
plot( NULL , xlim=c(-3,4) , ylim=c(0,0.35) ,
    xlab="log-odds survive" , ylab="Density" )

for ( i in 1:100 )
    curve( dnorm(x,post$a_bar[i],post$sigma[i]) , add=TRUE ,
    col=col.alpha("black",0.2) )

# sample 8000 imaginary tanks from the posterior distribution
sim_tanks <- rnorm( 8000 , post$a_bar , post$sigma )

# transform to probability and visualize
dens( inv_logit(sim_tanks) , lwd=2 , adj=0.1 )

## Varying effects and the underfitting/overfitting trade-off

### The model

<img src='pics/SR_r33.png' alt='Drawing' style='width:425pt'/>

### Assign values to the parameters

In [11]:
## R code 13.7
a_bar <- 1.5
sigma <- 1.5
nponds <- 60

# sample size N_i in each pond.
Ni <- as.integer( rep( c(5,10,25,35) , each=15 ) )


In [12]:
Ni

[1]  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5 10 10 10 10 10 10 10 10 10 10
[26] 10 10 10 10 10 25 25 25 25 25 25 25 25 25 25 25 25 25 25 25 35 35 35 35 35
[51] 35 35 35 35 35 35 35 35 35 35

The values $\bar{α}$ = 1.4 and σ = 1.5 define a Gaussian distribution of individual pond logodds of survival.

So now we need to simulate all 60 of these intercept values from the implied Gaussian distribution with mean $\bar{α}$ and standard deviation σ:

In [13]:
set.seed(5005)
a_pond <- rnorm( nponds , mean=a_bar , sd=sigma )

Go ahead and inspect the contents of a_pond. It should contain 60 log-odds values, one for each simulated pond.

In [14]:
a_pond %>% head()

[1]  0.5667312  1.9900232 -0.1377569  1.8567665  3.9120880  1.9541487

In [15]:
# let’s bundle some of this information in a data frame, 
# just to keep it organized.
dsim <- data.frame( pond=1:nponds , Ni=Ni , true_a=a_pond )

The first column is the pond index, 1 through 60. The second column is the initial tadpole count in each pond. The third column is the true log-odds survival for each pond.

In [16]:
dsim %>% head()

pond Ni true_a    
1 1    5   0.5667312
2 2    5   1.9900232
3 3    5  -0.1377569
4 4    5   1.8567665
5 5    5   3.9120880
6 6    5   1.9541487

### Simulate Survivors

<img src='pics/SR_r34.png' alt='Drawing' style='width:425pt'/>

**<font color = 'purple'>Putting the logistic into the random binomial function, we can generate a simulated survivor count for each pond: </font>**

In [17]:
# simulated survivor count
dsim$Si <- rbinom( nponds , prob=logistic(dsim$true_a) , size=dsim$Ni )

In [18]:
dsim %>% tail

pond Ni true_a    Si
55 55   35 0.5651822 26
56 56   35 2.5580613 33
57 57   35 0.5674268 23
58 58   35 2.7460641 35
59 59   35 1.5042225 29
60 60   35 2.4973795 32

As usual with R, if you give it a list of values, it returns a new list of the same length. 

<font color = 'green'>In the above, each paired $α_i$(`dsim$true_a`) and Ni (`dsim$Ni`) is used to generate a random survivor count with the appropriate probability of survival and maximum count. These counts are stored in a new column in `dsim$Si`</font>.

### Compute the no-pooling estimates

We’re ready to start analyzing the simulated data now. **<font color = 'purple' style='background-color: lightyellow'>The easiest task is to just compute the no-pooling estimates</font>**. 

* We can accomplish this straight from the empirical data, just by calculating the proportion of survivors in each pond. 

I’ll keep these estimates on the probability scale, instead of translating them to the log-odds scale, because we’ll want to compare the quality of the estimates on the probability scale later.

In [19]:
dsim$p_nopool <- dsim$Si / dsim$Ni

In [20]:
dsim$p_nopool %>% head()

[1] 0.8 1.0 0.2 1.0 1.0 1.0

<font color = 'blue'>Now there’s another column in dsim, containing the empirical proportions of survivors in each pond. 
    
*     These are the same no-pooling estimates you’d get by fitting a model with a dummy variable for each pond and flat priors that induce no regularization</font>.

### Compute the partial-pooling estimates

Now to fit the model to the simulated data, using map2stan. I’ll use a single long chain in this example, but keep in mind that you need to use multiple chains to check convergence to the right posterior distribution. In this case, it’s safe. But don’t get cocky.

In [21]:
## R code 13.13
dat <- list( Si=dsim$Si , Ni=dsim$Ni , pond=dsim$pond )
m13.3 <- ulam(
    alist(
        Si ~ dbinom( Ni , p ),
        logit(p) <- a_pond[pond],
        a_pond[pond] ~ dnorm( a_bar , sigma ),
        a_bar ~ dnorm( 0 , 1.5 ),
        sigma ~ dexp( 1 )
    ), data=dat , chains=4 )


Warning message in file(con, "w"):
"cannot open file 'C:\Users\bings\AppData\Local\Temp\Rtmp8YnaDH\file2efc373a6daa': No such file or directory"

ERROR: Error in file(con, "w"): cannot open the connection


We’ve fit the basic varying intercept model above. You can take a look at the estimates for α¯ and σ with the usual precis approach:

In [22]:
 precis( m13.3 , depth=2 ) %>% round(2)

ERROR: Error in precis(m13.3, depth = 2): object 'm13.3' not found


**<font color = 'purple' style='background-color: lightyellow'>Now let’s compute the predicted survival proportions and add those proportions to our growing simulation data frame</font>**. To indicate that it contains the partial pooling estimates, I’ll call the column `p_partpool`.

In [23]:
post <- extract.samples( m13.3 )
post$a_pond %>% dim()
post %>% length()
post %>% str

ERROR: Error in extract.samples(m13.3): object 'm13.3' not found


In [ ]:
post <- extract.samples( m13.3 )
dsim$p_partpool <- apply( inv_logit(post$a_pond) , 2 , mean )

In [ ]:
dsim %>% head(2)
'------'
post$a_pond %>% dim()
post$a_pond %>% head(2)

'------'

inv_logit(post$a_pond) %>% head(2) 

'--'

inv_logit(post$a_pond)[,1]  %>% mean()

If we want to compare to the **<font color = 'purple' style='background-color: lightyellow'>true per-pond survival probabilities</font>** used to generate the data, then we’ll also need to compute those, using the true_a column:

In [24]:
dsim$p_true <- inv_logit( dsim$true_a )

The last thing we need to do, before we can plot the results and realize the point of this lesson, is to **<font color = 'purple' style='background-color: lightyellow'>compute the absolute error between the estimates and the true varying effects</font>**. This is easy enough, using the existing columns:

In [25]:
nopool_error <- abs( dsim$p_nopool - dsim$p_true )
partpool_error <- abs( dsim$p_partpool - dsim$p_true )

In [26]:
options(repr.plot.width=7.59, repr.plot.height=4)

plot( 1:60 , nopool_error , xlab="pond" , ylab="absolute error" ,
col=rangi2 , pch=16 )
points( 1:60 , partpool_error )



ERROR: Error in xy.coords(x, y): 'x' and 'y' lengths differ


ERROR: Error in png(tf, width, height, "in", pointsize, bg, res, antialias = antialias): unable to start png() device


plot without title

## More than one type of cluster

### Multilevel chimpanzees

Here is the mathematical form of the model, with the new pieces of the machine highlighted in blue:
<img src='pics/SR_r35.png' alt='Drawing' style='width:425pt'/>

In [27]:
## R code 13.21
# library(rethinking)
data(chimpanzees)
d <- chimpanzees
d$treatment <- 1 + d$prosoc_left + 2*d$condition

dat_list <- list(
    pulled_left = d$pulled_left,
    actor = d$actor,
    block_id = d$block,
    treatment = as.integer(d$treatment) )

In [28]:
d$actor %>% str()

 int [1:504] 1 1 1 1 1 1 1 1 1 1 ...


In [4]:
set.seed(13)
m13.4 <- ulam(
    alist(
        pulled_left ~ dbinom( 1 , p ) ,
        logit(p) <- a[actor] + g[block_id] + b[treatment] ,
        b[treatment] ~ dnorm( 0 , 0.5 ),
        # adaptive priors
        a[actor] ~ dnorm( a_bar , sigma_a ),
        g[block_id] ~ dnorm( 0 , sigma_g ),
        # hyper-priors
        a_bar ~ dnorm( 0 , 1.5 ),
        sigma_a ~ dexp(1),
        sigma_g ~ dexp(1)
    ) , data=dat_list , chains=4 , cores=4 , log_lik=TRUE )

Warning message:
"There were 4 divergent transitions after warmup. Increasing adapt_delta above 0.95 may help. See
http://mc-stan.org/misc/warnings.html#divergent-transitions-after-warmup"Warning message:
"Examine the pairs() plot to diagnose sampling problems
"

In [ ]:
precis( m13.4 , depth=2 ) %>% round(2)
plot( precis(m13.4,depth=2) ) # also plot

You’ll end up with 2000 samples from 4 independent chains. As always, be sure to inspect the trace plots and the diagnostics. As soon as you start trusting the machine, the machine will betray your trust. In this case, you might see a warning about divergent transitions:

In [30]:
traceplot( m13.4 )

ERROR: Error in traceplot(m13.4): object 'm13.4' not found


As a consequence, adding block to this model hasn’t added a lot of overfitting risk. Let’s compare the model with only varying intercepts on actor to the model with both kinds of varying intercepts. **<font color = 'purple'>The model that ignores block is: </font>**

In [5]:
## R code 13.23
set.seed(14)
m13.5 <- ulam(
    alist(
        pulled_left ~ dbinom( 1 , p ) ,
        logit(p) <- a[actor] + b[treatment] ,
        b[treatment] ~ dnorm( 0 , 0.5 ),
        a[actor] ~ dnorm( a_bar , sigma_a ),
        a_bar ~ dnorm( 0 , 1.5 ),
        sigma_a ~ dexp(1)
    ) , data=dat_list , chains=4 , cores=4 , log_lik=TRUE )


In [7]:
m13.5 %>% stancode()

data{
    int block_id[504];
    int pulled_left[504];
    int treatment[504];
    int actor[504];
}
parameters{
    vector[4] b;
    vector[7] a;
    real a_bar;
    real<lower=0> sigma_a;
}
model{
    vector[504] p;
    sigma_a ~ exponential( 1 );
    a_bar ~ normal( 0 , 1.5 );
    a ~ normal( a_bar , sigma_a );
    b ~ normal( 0 , 0.5 );
    for ( i in 1:504 ) {
        p[i] = a[actor[i]] + b[treatment[i]];
        p[i] = inv_logit(p[i]);
    }
    pulled_left ~ binomial( 1 , p );
}
generated quantities{
    vector[504] log_lik;
    vector[504] p;
    for ( i in 1:504 ) {
        p[i] = a[actor[i]] + b[treatment[i]];
        p[i] = inv_logit(p[i]);
    }
    for ( i in 1:504 ) log_lik[i] = binomial_lpmf( pulled_left[i] | 1 , p[i] );
}



**<font color = 'purple'>Comparing to the model with both clusters: </font>**

In [6]:
compare( m13.4 , m13.5 ) %>% round(2)

WAIC   pWAIC dWAIC weight SE    dSE
m13.5 530.96  8.45 0.00  0.65   19.23  NA
m13.4 532.24 10.58 1.28  0.35   19.35 1.6

**<font color = 'purple'>While m13.4 has 7 more parameters than m13.5 does, it has only about 2.5 more effective parameters. Why? </font>** 

* Because the posterior distribution for sigma_g ended up close to zero. This means each of the 6 g parameters is strongly shrunk towards zero—they are relatively inflexible. 


* In contrast, the a parameters are shrunk towards zero much less, because the estimated variation across actors is much larger, resulting in less shrinkage. <font color = 'blue'>But as a consequence, each of the a parameters contributes much more to the pWAIC value</font>.

**<font color = 'purple'>You might also notice that the difference in WAIC between these models is small </font>**, only 2. This is especially small compared the standard deviation of the difference, 1.6. These two models imply nearly identical predictions, and so their expected out-of-sample accuracy is nearly identical. <font color = 'blue'>The block parameters have been shrunk so much towards zero that they do very little work in the model.</font>

### Even more clusters

In [2]:
## R code 13.21
library(rethinking)
data(chimpanzees)
d <- chimpanzees
d$treatment <- 1 + d$prosoc_left + 2*d$condition

dat_list <- list(
    pulled_left = d$pulled_left,
    actor = d$actor,
    block_id = d$block,
    treatment = as.integer(d$treatment) )

In [ ]:
## R code 13.25
set.seed(15)
m13.6 <- ulam(
    alist(
        pulled_left ~ dbinom( 1 , p ) ,
        logit(p) <- a[actor] + g[block_id] + b[treatment] ,
        b[treatment] ~ dnorm( 0 , sigma_b ),
        a[actor] ~ dnorm( a_bar , sigma_a ),
        g[block_id] ~ dnorm( 0 , sigma_g ),
        a_bar ~ dnorm( 0 , 1.5 ),
        sigma_a ~ dexp(1),
        sigma_g ~ dexp(1),
        sigma_b ~ dexp(1)
    ) , data=dat_list , chains=4 , cores=4 , log_lik=TRUE )

In [3]:
coeftab(m13.4,m13.6)

Warning message:
"There were 7 divergent transitions after warmup. Increasing adapt_delta above 0.95 may help. See
http://mc-stan.org/misc/warnings.html#divergent-transitions-after-warmup"Warning message:
"Examine the pairs() plot to diagnose sampling problems
"

ERROR: Error in coeftab(m13.4, m13.6): object 'm13.4' not found


## Divergent transitions and non-centered priors

### Adjust target acceptance

In [34]:
divergent(m13.4)

ERROR: Error in class(fit) %in% c("map2stan", "ulam"): object 'm13.4' not found


In [ ]:
set.seed(13)
m13.4b <- ulam( m13.4 , chains=4 , cores=4 , control=list(adapt_delta=0.99) )
divergent(m13.4b)

In [ ]:
m13.4b %>% precis()

### Reparameterization

In [35]:
## R code 13.28
set.seed(13)
m13.4nc <- ulam(
    alist(
        pulled_left ~ dbinom( 1 , p ) ,
        logit(p) <- a_bar + z[actor]*sigma_a + x[block_id]*sigma_g + b[treatment] ,
        b[treatment] ~ dnorm( 0 , 0.5 ),
        z[actor] ~ dnorm( 0 , 1 ),
        x[block_id] ~ dnorm( 0 , 1 ),
        a_bar ~ dnorm( 0 , 1.5 ),
        sigma_a ~ dexp(1),
        sigma_g ~ dexp(1)
    ) , data=dat_list , chains=4 , cores=4 )

Warning message in file(con, "w"):
"cannot open file 'C:\Users\bings\AppData\Local\Temp\Rtmp8YnaDH\file2efc16573b1a': No such file or directory"

ERROR: Error in file(con, "w"): cannot open the connection


In [ ]:
## R code 13.29
neff_c <- precis( m13.4 , depth=2 )[['n_eff']]
neff_nc <- precis( m13.4nc , depth=2 )[['n_eff']]
par_names <- rownames( precis( m13.4 , depth=2 ) )
neff_table <- cbind( neff_c , neff_nc )
rownames(neff_table) <- par_names
neff_table %>% t() %>% round()


## Multilevel posterior predictions

**<font color = 'purple' style='background-color: lightyellow'>We’ll construct posterior predictions (retrodictions), using both the automated link approach and doing it from scratch, so there is no confusion.</font>**

### Posterior prediction for same clusters.

The code needed to compute posterior predictions is just like the code from Chapter 11. Here it is again, **<font color = 'purple'>computing and plotting posterior predictions for actor number 2: </font>**

##### Link version

In [36]:
## R code 13.30
chimp <- 2
d_pred <- list(
    actor = rep(chimp,4),
    treatment = 1:4,
    block_id = rep(1,4)
)
p <- link( m13.4 , data=d_pred )
p_mu <- apply( p , 2 , mean )
p_ci <- apply( p , 2 , PI )


ERROR: Error in link(m13.4, data = d_pred): object 'm13.4' not found


##### non-link version

**<font color = 'purple'>To construct the same calculations without using link </font>**, we just have to remember the model. The only difficulty is that when we work with the samples from the posterior, the varying intercepts will be a matrix of samples. Let’s take a look:

In [37]:
post <- extract.samples(m13.4)
str(post)

ERROR: Error in extract.samples(m13.4): object 'm13.4' not found


In [ ]:
post$a %>% head(2)

post$b %>% head(4)

The a matrix has samples on the rows and actors on the columns. So to plot, for example, the density for actor 5:

In [38]:
dens( post$a[,5] )

ERROR: Error in dens(post$a[, 5]): object 'post' not found


The `[,5]` means “all samples for actor 5.”

**<font color = 'purple'>To construct posterior predictions, we build our own link function. </font>** 

<font color = 'blue'>I’ll use the with function here, so we don’t have to keep typing post$ before every parameter name:</font>

In [39]:
## R code 13.33
p_link <- function( treatment , actor=1 , block_id=1 ) {
    logodds <- with( post , a[,actor] + g[,block_id] + b[,treatment] )
    return( inv_logit(logodds) )
}


`logodds <- post$a[,actor] + post$g[,block_id] + post$b[,treatment]`

same meaning as the `with` statement line

The linear model is identical to the one used to define the model, but with a single comma added inside the brackets after a. Now to compute predictions:

In [40]:
## R code 13.34
p_raw <- sapply( 1:4 , function(i) p_link( i , actor=2 , block_id=1 ) )
p_mu <- apply( p_raw , 2 , mean )
p_ci <- apply( p_raw , 2 , PI )

ERROR: Error in with(post, a[, actor] + g[, block_id] + b[, treatment]): object 'post' not found


At some point, you will have to work with a model that link will mangle. At that time, you can return to this section and peer hard at the code above and still make progress. No matter what the model is, if it is a Bayesian model, then it is generative. This means that predictions are made by pushing samples up through the model to get distributions of predictions. Then you summarize the distributions to summarize the predictions.


### Posterior prediction for new clusters.

**<font color = 'purple'>let’s see how to construct posterior predictions for a new, previously unobserved average actor </font>**. 

* By “average,” I mean an individual chimpanzee with an intercept exactly at a_bar ($\bar{α}$), the population mean. Since there is uncertainty about the population mean, there is still uncertainty about this average individual’s intercept. But as you’ll see, the uncertainty is much smaller than it really should be, if we wish to honestly represent the problem of what to expect from a new individual. 

What we need is our own link function, but now with twist:

In [41]:
## R code 13.35
p_link_abar <- function( treatment ) {
    logodds <- with( post , a_bar + b[,treatment] )
    return( inv_logit(logodds) )
}


Notice that the function ignores block. This is because we are extrapolating to new blocks,
so we assume the average block effect is about zero (which it was in the sample). Call this
function and summarize just as before:

In [42]:

## R code 13.36
p_raw <- sapply( 1:4 , function(i) p_link_abar( i ) )
p_mu <- apply( p_raw , 2 , mean )
p_ci <- apply( p_raw , 2 , PI )

plot( NULL , xlab="treatment" , ylab="proportion pulled left" ,
    ylim=c(0,1) , xaxt="n" , xlim=c(1,4) )
axis( 1 , at=1:4 , labels=c("R/N","L/N","R/P","L/P") )
lines( 1:4 , p_mu )
shade( p_ci , 1:4 )


ERROR: Error in with(post, a_bar + b[, treatment]): object 'post' not found


The gray region shows the 80% compatibility interval for an actor with an average intercept. This kind of calculation makes it easy to see the impact of prosoc_left, as well as uncertainty about where the average is, but it doesn’t show the variation among actors.

To show the variation among actors, we’ll need to use sigma_a in the calculation. First we simply use rnorm to sample some random chimpanzees, using mean a_bar and standard deviation sigma_a. Then we write a link function that references those simulated chimpanzees, not the ones in the posterior. 

* It’s important to do the chimpanzee sampling outside the link function, because we want to reference the same simulate chimpanzee, whichever treatment we consider. 

This is the code:

In [43]:

## R code 13.37
a_sim <- with( post , rnorm( length(post$a_bar) , a_bar , sigma_a ) )
p_link_asim <- function( treatment ) {
    logodds <- with( post , a_sim + b[,treatment] )
    return( inv_logit(logodds) )
}
p_raw_asim <- sapply( 1:4 , function(i) p_link_asim( i ) )

ERROR: Error in with(post, rnorm(length(post$a_bar), a_bar, sigma_a)): object 'post' not found


In [ ]:
## R code 13.38
plot( NULL , xlab="treatment" , ylab="proportion pulled left" ,
    ylim=c(0,1) , xaxt="n" , xlim=c(1,4) )
axis( 1 , at=1:4 , labels=c("R/N","L/N","R/P","L/P") )
for ( i in 1:100 ) lines( 1:4 , p_raw_asim[i,] , col=col.alpha("black",0.25) , lwd=2 )


## Practice Problems

In [2]:
library(tidyverse)

-- Attaching packages --------------------------------------- tidyverse 1.2.1 --
v tibble  2.0.1       v purrr   0.3.0  
v tidyr   0.8.2       v dplyr   0.8.0.1
v readr   1.3.1       v stringr 1.4.0  
v tibble  2.0.1       v forcats 0.4.0  
Warning message:
"package 'tibble' was built under R version 3.5.2"Warning message:
"package 'readr' was built under R version 3.5.2"Warning message:
"package 'purrr' was built under R version 3.5.2"Warning message:
"package 'dplyr' was built under R version 3.5.2"Warning message:
"package 'stringr' was built under R version 3.5.2"Warning message:
"package 'forcats' was built under R version 3.5.2"-- Conflicts ------------------------------------------ tidyverse_conflicts() --
x tidyr::extract() masks rstan::extract()
x dplyr::filter()  masks stats::filter()
x dplyr::lag()     masks stats::lag()
x purrr::map()     masks rethinking::map()


In [ ]:
library(rethinking)
data(reedfrogs)
d <- reedfrogs

In [4]:
str(d)
d %>% head

'data.frame':	48 obs. of  5 variables:
 $ density : int  10 10 10 10 10 10 10 10 10 10 ...
 $ pred    : Factor w/ 2 levels "no","pred": 1 1 1 1 1 1 1 1 2 2 ...
 $ size    : Factor w/ 2 levels "big","small": 1 1 1 1 2 2 2 2 1 1 ...
 $ surv    : int  9 10 7 10 9 9 10 9 4 9 ...
 $ propsurv: num  0.9 1 0.7 1 0.9 0.9 1 0.9 0.4 0.9 ...


density pred size  surv propsurv
1 10      no   big    9   0.9     
2 10      no   big   10   1.0     
3 10      no   big    7   0.7     
4 10      no   big   10   1.0     
5 10      no   small  9   0.9     
6 10      no   small  9   0.9

In [7]:
d$pred <- ifelse( d$pred=="no" , 0 , 1 )
d$big <- ifelse( d$size=="big" , 1 , 0 )
d$tank <- 1:nrow(d)

In [8]:
m0 <- map2stan(
    alist(
        surv ~ dbinom(density,p),
        logit(p) <- a_tank[tank],
        a_tank[tank] ~ dnorm(a,sigma_tank),
        a ~ dnorm(0,10),
        sigma_tank ~ dcauchy(0,1)
    ),
    data=d , chains=4 )


SAMPLING FOR MODEL '134938cb467d54dc5d6211d037aee09f' NOW (CHAIN 1).
Chain 1: 
Chain 1: Gradient evaluation took 0 seconds
Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 0 seconds.
Chain 1: Adjust your expectations accordingly!
Chain 1: 
Chain 1: 
Chain 1: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 1: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 1: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 1: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 1: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 1: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 1: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 1: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 1: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 1: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 1: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 1: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 1: 
Chain 1:  Elapsed Time: 0.446 seconds (Warm-up)
Chain 1:                0.526 seconds (Sampling)
Chain 1:       

Computing WAIC


In [10]:
m_p <- map2stan( 
    alist(
        surv ~ dbinom(density,p),
        logit(p) <- a_tank[tank] + bp*pred,
        bp ~ dnorm(0,1),
        a_tank[tank] ~ dnorm(a,sigma_tank),
        a ~ dnorm(0,10),
        sigma_tank ~ dcauchy(0,1)
    ),
    data=d , chains=4 )


SAMPLING FOR MODEL '8ee51b6da32fa6dcc1741095e3c2b4fd' NOW (CHAIN 1).
Chain 1: 
Chain 1: Gradient evaluation took 0 seconds
Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 0 seconds.
Chain 1: Adjust your expectations accordingly!
Chain 1: 
Chain 1: 
Chain 1: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 1: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 1: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 1: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 1: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 1: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 1: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 1: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 1: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 1: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 1: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 1: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 1: 
Chain 1:  Elapsed Time: 2.019 seconds (Warm-up)
Chain 1:                1.892 seconds (Sampling)
Chain 1:       

Computing WAIC


In [11]:
m_b <- map2stan( 
    alist(
        surv ~ dbinom(density,p),
        logit(p) <- a_tank[tank] + bb*big,
        bb ~ dnorm(0,1),
        a_tank[tank] ~ dnorm(a,sigma_tank),
        a ~ dnorm(0,10),
        sigma_tank ~ dcauchy(0,1)
    ),
    data=d , chains=4 )


SAMPLING FOR MODEL 'de7f0492e5174e0ea6dd9848e77b71ec' NOW (CHAIN 1).
Chain 1: 
Chain 1: Gradient evaluation took 0 seconds
Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 0 seconds.
Chain 1: Adjust your expectations accordingly!
Chain 1: 
Chain 1: 
Chain 1: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 1: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 1: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 1: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 1: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 1: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 1: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 1: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 1: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 1: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 1: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 1: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 1: 
Chain 1:  Elapsed Time: 0.585 seconds (Warm-up)
Chain 1:                0.754 seconds (Sampling)
Chain 1:       

Computing WAIC


In [26]:
m_p_b <- map2stan( 
    alist(
        surv ~ dbinom(density,p),
        logit(p) <- a_tank[tank] + bp*pred + bb*big,
        c(bp,bb) ~ dnorm(0,1),
        a_tank[tank] ~ dnorm(a,sigma_tank),
        a ~ dnorm(0,10),
        sigma_tank ~ dcauchy(0,1)
    ),
    data=d , chains=4 )


SAMPLING FOR MODEL 'e19577fc1514bee1955b27ed9ba845ec' NOW (CHAIN 1).
Chain 1: 
Chain 1: Gradient evaluation took 0 seconds
Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 0 seconds.
Chain 1: Adjust your expectations accordingly!
Chain 1: 
Chain 1: 
Chain 1: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 1: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 1: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 1: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 1: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 1: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 1: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 1: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 1: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 1: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 1: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 1: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 1: 
Chain 1:  Elapsed Time: 1.011 seconds (Warm-up)
Chain 1:                0.948 seconds (Sampling)
Chain 1:       

Computing WAIC


In [24]:
m_p_b_pb <- map2stan(
    alist(
        surv ~ dbinom(density,p),
        logit(p) <- a_tank[tank] + bp*pred + bb*big + bpb*pred*big,
        c(bp,bb,bpb) ~ dnorm(0,1),
        a_tank[tank] ~ dnorm(a,sigma_tank),
        a ~ dnorm(0,10),
        sigma_tank ~ dcauchy(0,1)
    ),
    data=d , chains=4 )


SAMPLING FOR MODEL '4127bc9e8f7e25c7e77b1c35b368d93e' NOW (CHAIN 1).
Chain 1: 
Chain 1: Gradient evaluation took 0 seconds
Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 0 seconds.
Chain 1: Adjust your expectations accordingly!
Chain 1: 
Chain 1: 
Chain 1: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 1: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 1: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 1: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 1: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 1: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 1: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 1: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 1: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 1: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 1: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 1: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 1: 
Chain 1:  Elapsed Time: 1.431 seconds (Warm-up)
Chain 1:                1.464 seconds (Sampling)
Chain 1:       

Computing WAIC


In [19]:
 coeftab(m0,m_p,m_b,m_p_b,m_p_b_pb)

           m0      m_p     m_b     m_p_b   m_p_b_pb
a_tank[1]     2.15    2.00    2.49    2.30    2.39 
a_tank[2]     3.11    2.94    3.35    3.25    3.28 
a_tank[3]     1.00    0.84    1.34    1.20    1.25 
a_tank[4]     3.09    2.96    3.36    3.22    3.31 
a_tank[5]     2.15    2.00    2.22    2.07    2.11 
a_tank[6]     2.17    1.99    2.19    2.07    2.11 
a_tank[7]     3.10    2.93    3.19    3.05    3.11 
a_tank[8]     2.17    2.00    2.20    2.06    2.14 
a_tank[9]    -0.17   -0.33    0.17    0.02    0.09 
a_tank[10]    2.15    1.97    2.47    2.31    2.40 
a_tank[11]    1.01    0.85    1.35    1.20    1.23 
a_tank[12]    0.59    0.44    0.94    0.77    0.84 
a_tank[13]    1.01    0.86    1.05    0.91    0.97 
a_tank[14]    0.20    0.03    0.23    0.11    0.14 
a_tank[15]    2.14    1.98    2.18    2.08    2.13 
a_tank[16]    2.12    1.98    2.20    2.09    2.13 
a_tank[17]    2.92    2.76    3.24    3.11    3.15 
a_tank[18]    2.39    2.24    2.73    2.59    2.64 
a_tank[19]  

In [21]:
precis(m_p_b_pb) %>% round(2)

48 vector or matrix parameters hidden. Use depth=2 to show them.


mean  sd   5.5%  94.5% n_eff   Rhat
bp          0.08 0.95 -1.48 1.57   153.15 1.02
bb         -0.19 0.74 -1.36 1.00  1302.01 1.00
bpb        -0.17 0.74 -1.34 0.99  1246.70 1.00
a           1.48 1.00 -0.09 3.11   164.11 1.02
sigma_tank  1.63 0.23  1.30 2.01  1288.07 1.00

# ---
<hr>

Use extract.samples and extract.prior to return lists of samples for posterior and prior distributions, respectively.

In [44]:
# dat_list %>% head
dat_list$pulled_left %>% length
dat_list$pulled_left %>% length
p %>% head
p %>% dim

[1] 504

[1] 504

ERROR: Error in eval(lhs, parent, parent): object 'p' not found


In [ ]:
## R code 13.21
# library(rethinking)
data(chimpanzees)
d <- chimpanzees
d$treatment <- 1 + d$prosoc_left + 2*d$condition

dat_list <- list(
    pulled_left = d$pulled_left,
    actor = d$actor,
    block_id = d$block,
    treatment = as.integer(d$treatment) )

# Misc

##### With function

In [45]:
with(mtcars, mpg[cyl == 8  &  disp > 350])
    # is the same as, but nicer than

mtcars$mpg[mtcars$cyl == 8  &  mtcars$disp > 350]

[1] 18.7 14.3 10.4 10.4 14.7 19.2 15.8

[1] 18.7 14.3 10.4 10.4 14.7 19.2 15.8

There are many functions that work on data frames and take a data argument so that you don't need to retype the name of the data frame for every time you reference a column. lm, plot.formula, subset, transform are just a few examples.

with is a general purpose wrapper to let you use any function as if it had a data argument.

Using the mtcars data set, we could fit a model with or without using the data argument:

In [46]:
# this is obviously annoying
mod = lm(mtcars$mpg ~ mtcars$cyl + mtcars$disp + mtcars$wt)

# this is nicer
mod = lm(mpg ~ cyl + disp + wt, data = mtcars)

However, if (for some strange reason) we wanted to find the mean of cyl + disp + wt, there is a problem because mean doesn't have a data argument like lm does. This is the issue that `with` addresses:

In [47]:
# without with(), we would be stuck here:
z = mean(mtcars$cyl + mtcars$disp + mtcars$wt)

# using with(), we can clean this up:
z = with(mtcars, mean(cyl + disp + wt))

##### Overthinking: Repeating the pond simulation.

**Overthinking**: Repeating the pond simulation. This model samples pretty quickly. Compiling the model takes up most of the execution time. Luckily the compilation only has to be done once. Then you can pass new data to the compiled model and get new estimates. Once you’ve compiled m13.3 once, you can use this code to re-simulate ponds and sample from the new posterior, without waiting for the model to compile again:

In [48]:
a <- 1.5
sigma <- 1.5
nponds <- 60
Ni <- as.integer( rep( c(5,10,25,35) , each=15 ) )
a_pond <- rnorm( nponds , mean=a , sd=sigma )
dsim <- data.frame( pond=1:nponds , Ni=Ni , true_a=a_pond )
dsim$Si <- rbinom( nponds,prob=inv_logit( dsim$true_a ),size=dsim$Ni )
dsim$p_nopool <- dsim$Si / dsim$Ni
newdat <- list(Si=dsim$Si,Ni=dsim$Ni,pond=1:nponds)
m13.3new <- stan( fit=m13.3@stanfit , data=newdat , chains=4 )

post <- extract.samples( m13.3new )
dsim$p_partpool <- apply( inv_logit(post$a_pond) , 2 , mean )
dsim$p_true <- inv_logit( dsim$true_a )
nopool_error <- abs( dsim$p_nopool - dsim$p_true )
partpool_error <- abs( dsim$p_partpool - dsim$p_true )
plot( 1:60 , nopool_error , xlab="pond" , ylab="absolute error" , col=rangi2 , pch=16 )
points( 1:60 , partpool_error )

ERROR: Error in is(fit, "stanfit"): object 'm13.3' not found


##### rnorm

In [49]:
rnorm( 10 , mean = 1.5 , sd=1.5 )

[1]  2.0698436  0.6517420 -0.3206671 -0.5464524 -0.6241994  1.1166330
 [7] -0.3381389  1.8207514  1.6008353  2.7849527

##### log odds

see chapter 11

* A unit change in xi may produce a larger or smaller change in the probability pi, depending upon how far from zero the log-odds are. 

For example, in Figure 10.7, <font color = 'blue'> when x = 0 the linear model has a value of zero on the log-odds scale. A halfunit increase in x results in about a 0.25 increase in probability. But each addition half-unit will produce less and less of an increase in probability, until any increase is vanishingly small.</font> 

* And if you think about it, a good model of probability needs to behave this way. When an event is almost guaranteed to happen, its probability cannot increase very much, no matter how important the predictor may be.

<img src='pics/Snap32.png' alt='Drawing' style='width:425pt'/>

##### Probabilities from density curves

https://www.khanacademy.org/math/statistics-probability/random-variables-stats-library/random-variables-continuous/v/probabilities-from-density-curves

##### Curve Function 

https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/curve.html

*The function or expression expr (for curve) or function x (for plot) is evaluated at n points equally spaced over the range [from, to]. The points determined in this way are then plotted.

* If either from or to is NULL, it defaults to the corresponding element of xlim if that is not NULL.

* For curve(add = NA) and curve(add = TRUE) the defaults are taken from the x-limits used for the previous plot.

##### Why QUAP fails, MCMC succeeds.

**Overthinking**: QUAP fails, MCMC succeeds. Why doesn’t simple quadratic approximation, using for example quap, work with multilevel models? When a prior is itself a function of parameters, there are two levels of uncertainty. This means that the probability of the data, conditional on the parameters, must average over each level. Ordinary quadratic approximation cannot handle the averaging in the likelihood, because in general it’s not possible to derive an analytical solution. That means there is no unified function for calculating the log-posterior. So your computer cannot directly find its minimum (the maximum of the posterior). Some other computational approach is needed. It is possible to extend the mode-finding optimization strategy to these models, but we don’t want to be stuck with optimization in general. One reason is that the posterior of these models is routinely non-Gaussian. More generally, as models become more complex, a phenomenon known as concentration of measure guarantees that the posterior mode will be far from the posterior median. So we really need to give up optimization as a strategy. One robust solution is MCMC.

A Gaussian approximation is called “quadratic approximation” because the logarithm of a Gaussian distribution forms a parabola. And a parabola is a quadratic function. So this approximation essentially represents any log-posterior with a parabola.

For many of the most common procedures in applied statistics—linear regression, for example—the approximation works very well.

The procedure, which R will happily conduct (QUAC) at your command, contains two steps.

* (1) Find the posterior mode. This is usually accomplished by some optimization algorithm, a procedure that virtually “climbs” the posterior distribution, as if it were a mountain. The golem doesn’t know where the peak is, but it does know the slope under its feet. There are many well-developed optimization procedures, most of them more clever than simple hill climbing. But all of them try to find peaks. 


* (2) Once you find the peak of the posterior, you must estimate the curvature near the peak. This curvature is sufficient to compute a quadratic approximation of the entire posterior distribution. In some cases, these calculations can be done analytically, but usually your computer uses some numerical technique instead.

##### Regularizing Priors

Our job is to carefully navigate between overfitting and underfitting. There are two common families of approaches. 

* The first approach is to use a <span style='background-color: lightgreen'>regularizing prior</span> to <font color = 'blue'>tell the model not to get too excited by the data</font>. This is the same device that non-Bayesian methods refer to as “penalized likelihood.” 


* The second approach is to use some scoring device, like <span style='background-color: lightgreen'>information criteria</span>, to model the prediction task and estimate predictive accuracy for some purpose. Both families of approaches are routinely used in the natural and social sciences. Furthermore, they can be—maybe should be—used in combination. So it’s worth understanding both, as you’re going to need both at some point.

## How link works

<img src='pics/SR_r37.png' alt='Drawing' style='width:425pt'/>

##### Link example M4.3

What link will do is take your quap approximation, sample from the posterior distribution, and then compute µ for each case in the data and sample from the posterior distribution.

In [50]:
## R code 4.42
# load data again, since it's a long way back
library(rethinking)
data(Howell1)
d <- Howell1
d2 <- d[ d$age >= 18 , ]

# define the average weight, x-bar
xbar <- mean(d2$weight)

# fit model
m4.3 <- quap(
    alist(
        height ~ dnorm( mu , sigma ) ,
        mu <- a + b*( weight - xbar ) ,
        a ~ dnorm( 178 , 20 ) ,
        b ~ dlnorm( 0 , 1 ) ,
        sigma ~ dunif( 0 , 50 )
    ) ,
    data=d2 )

In [51]:
d2 %>% head
d2 %>% dim

height  weight   age male
1 151.765 47.82561 63  1   
2 139.700 36.48581 63  0   
3 136.525 31.86484 65  0   
4 156.845 53.04191 41  1   
5 145.415 41.27687 51  0   
6 163.830 62.99259 35  1

[1] 352   4

In [52]:
mu <- link( m4.3 ) 
mu %>% head()

mu %>% dim

[,1]     [,2]     [,3]     [,4]     [,5]     [,6]     [,7]     [,8]    
[1,] 157.4431 146.6290 142.2223 162.4176 151.1980 171.9070 148.3052 164.7426
[2,] 157.6083 146.6297 142.1560 162.6584 151.2682 172.2921 148.3314 165.0188
[3,] 157.7038 147.9141 143.9248 162.2071 152.0502 170.7975 149.4315 164.3119
[4,] 157.2676 146.1495 141.6188 162.3820 150.8469 172.1382 147.8728 164.7724
[5,] 156.6896 145.8559 141.4411 161.6732 150.4331 171.1798 147.5351 164.0024
[6,] 156.6643 146.9399 142.9773 161.1375 151.0485 169.6706 148.4472 163.2282
     [,9]     [,10]    [,11] [,12]    [,13]    [,14]    [,15]    [,16]   
[1,] 145.0880 163.7964 ...   153.8204 157.4972 149.4137 150.9817 150.7654
[2,] 145.0653 164.0582 ...   153.9305 157.6632 149.4567 151.0486 150.8290
[3,] 146.5191 163.4553 ...   154.4243 157.7528 150.4349 151.8545 151.6587
[4,] 144.5651 163.7996 ...   153.5430 157.3232 149.0124 150.6245 150.4022
[5,] 144.3121 163.0545 ...   153.0603 156.7438 148.6456 150.2165 149.9998
[6,] 145.5542 162.3773 ...   153.4066 156.7129 149.4440 150.8540 150.6595
     [,17]    [,18]    [,19]    [,20]    [,21]   
[1,] 156.6861 144.4933 161.5795 163.3909 161.9310
[2,] 156.8398 144.4615 161.8076 163.6465 162.1644
[3,] 157.0185 145.9806 161.4484 163.0881 161.7665
[4,] 156.4894 143.9536 161.5203 163.3826 161.8817
[5,] 155.9313 143.7162 160.8335 162.6482 161.1856
[6,] 155.9836 145.0194 160.3838 162.0126 160.6999

[1] 1000  352

You end up with a big matrix of values of µ. 

* <font color = 'green'>Each row is a sample from the posterior distribution</font>. The default is 1000 samples, but you can use as many or as few as you like. 


* <font color = 'green'>Each column is a case (row) in the data; i.e., a value of weight</font>. There are 352 rows in d2, corresponding to 352 individuals. So there are 352 columns in the matrix mu above.

##### **Overthinking**: How link works.

The function link is not really very sophisticated. 

* All it is doing is using the formula you provided when you fit the model to compute the value of the linear model. It does this for each sample from the posterior distribution, for each case in the data. 

You could accomplish the same thing for any model, fit by any means, by performing these steps yourself. This is how it’d look for m4.3.

In [53]:
post <- extract.samples(m4.3)
mu.link <- function(weight) post$a + post$b*( weight - xbar )
weight.seq <- seq( from=25 , to=70 , by=1 )
mu <- sapply( weight.seq , mu.link )
mu.mean <- apply( mu , 2 , mean )
mu.HPDI <- apply( mu , 2 , HPDI , prob=0.89 )

In [54]:
## R code 4.55
# use type="n" to hide raw data
plot( height ~ weight , d2 , type="n" )

# loop over samples and plot each mu value
for ( i in 1:100 )
    points( weight.seq , mu[i,] , pch=16 , col=col.alpha(rangi2,0.1) )

options(repr.plot.width=6, repr.plot.height=5)

ERROR: Error in png(tf, width, height, "in", pointsize, bg, res, antialias = antialias): unable to start png() device


plot without title

In [ ]:
## R code 4.56
# summarize the distribution of mu
mu.mean <- apply( mu , 2 , mean )
mu.HPDI <- apply( mu , 2 , HPDI , prob=0.89 )

## R code 4.57
# plot raw data
# fading out points to make line and interval more visible
plot( height ~ weight , data=d2 , col=col.alpha(rangi2,0.5) )

# plot the MAP line, aka the mean mu for each weight
lines( weight.seq , mu.mean )

# plot a shaded region for 89% HPDI
shade( mu.HPDI , weight.seq )



<img src='pics/SR_r38.png' alt='Drawing' style='width:425pt'/>

mu.mean 

In [55]:
mu %>% dim

[1] 10000    46

And the values in mu.mean and mu.HPDI should be very similar (allowing for simulation variance) to what you got the automated way, using link. 

Knowing this manual method is useful both for (1) understanding and (2) sheer power. Whatever the model you find yourself with, this approach can be used to generate posterior predictions for any component of it. Automated tools like link save effort, but they are never as flexible as the code you can write yourself.

In [56]:
mu %>% head()

[,1]     [,2]     [,3]     [,4]     [,5]     [,6]     [,7]     [,8]    
[1,] 135.4750 136.4390 137.4031 138.3671 139.3311 140.2951 141.2592 142.2232
[2,] 136.3289 137.2285 138.1281 139.0277 139.9272 140.8268 141.7264 142.6260
[3,] 136.4446 137.3696 138.2946 139.2196 140.1446 141.0696 141.9946 142.9196
[4,] 135.8260 136.7810 137.7359 138.6909 139.6458 140.6008 141.5557 142.5107
[5,] 135.8642 136.8014 137.7387 138.6759 139.6132 140.5505 141.4877 142.4250
[6,] 137.6323 138.5035 139.3747 140.2459 141.1170 141.9882 142.8594 143.7306
     [,9]     [,10]    [,11] [,12]    [,13]    [,14]    [,15]    [,16]   
[1,] 143.1872 144.1513 ...   170.1801 171.1441 172.1081 173.0721 174.0362
[2,] 143.5255 144.4251 ...   168.7137 169.6132 170.5128 171.4124 172.3120
[3,] 143.8447 144.7697 ...   169.7448 170.6698 171.5948 172.5198 173.4448
[4,] 143.4656 144.4205 ...   170.2041 171.1590 172.1139 173.0689 174.0238
[5,] 143.3622 144.2995 ...   169.6055 170.5428 171.4800 172.4173 173.3545
[6,] 144.6018 145.4730 ...   168.9949 169.8661 170.7373 171.6085 172.4797
     [,17]    [,18]    [,19]    [,20]    [,21]   
[1,] 175.0002 175.9642 176.9283 177.8923 178.8563
[2,] 173.2115 174.1111 175.0107 175.9103 176.8098
[3,] 174.3698 175.2948 176.2198 177.1448 178.0698
[4,] 174.9788 175.9337 176.8887 177.8436 178.7986
[5,] 174.2918 175.2290 176.1663 177.1036 178.0408
[6,] 173.3508 174.2220 175.0932 175.9644 176.8356

###### sapply(X, FUN, ...)

<font color = 'blue'>takes each value in the first argument and passes it into the function in the second argument</font>

X = a vector (atomic or list) or an expression object. Other objects (including classed objects) will be coerced by base::as.list.

FUN = the function to be applied to each element of X: see ‘Details’. In the case of functions like +, %*%, the function name must be backquoted or quoted.

###### apply(X, MARGIN, FUN, ...)

**<font color = 'purple'>X </font>**	
an array, including a matrix.

**<font color = 'purple'>MARGIN </font>**	
a vector giving the subscripts which the function will be applied over. E.g., for a matrix 1 indicates rows, 2 indicates columns, c(1, 2) indicates rows and columns. Where X has named dimnames, it can be a character vector selecting dimension names.

**<font color = 'purple'>FUN </font>**	
the function to be applied: see ‘Details’. In the case of functions like +, %*%, etc., the function name must be backquoted or quoted.